### ISSUE TITLE
snmp timeouts when using cmd.php

### ISSUE THREAD

**Issue by seppeel** [Cacti/cacti#1498]
Hi,
im having issues running the poller on cacti 1.1.37
The following error is generated only sometimes and the snmp query which fails is changing:

```
27.03.2018 09:30:16 - POLLER: Poller[Main Poller] Device[57] Device[CP-VSX - FW-ZWO] Graphs[CP-VSX - FW-ZWO - Traffic - 10.21.1.113 (wrp1047), CP-VSX - FW-ZWO - Traffic - 10.21.1.113 (wrp1047)] DS[CP-VSX - FW-ZWO - Traffic - 10.21.1.113 - wrp1047] WARNING: Result from SNMP not valid.  Partial Result: U
--
27.03.2018 09:30:16 - CMDPHP PHP ERROR WARNING Backtrace: (/cmd.php: 659 cacti_snmp_session_get)(/lib/snmp.php: 472 get)(CactiErrorHandler)(/lib/functions.php: 4585 cacti_debug_backtrace)
27.03.2018 09:30:16 - ERROR PHP WARNING: SNMP::get(): Fatal error: Timeout in file: /var/www/cacti/lib/snmp.php  on line: 472
```

There are gaps in the graphs if these logs are generated.
i already increased the timeout for snmp queries.
im running an old cacti server (0.8.8) parallel. The old cacti server does not have these timeouts.
stopping the poller on the old cacti server doesnt solve this problem, too (was thinking of too much snmp spam in the network).

Any ideas regarding this problem? thanks for your efforts :)
seppel

**Comment by netniV** [comment]
There is an option somewhere I think that says to use the last value, I think it's snmp retries or an option near there.  If you can't spot it I'll take a look later.

**Comment by seppeel** [comment]
i didnt find that option, sorry.

**Comment by netniV** [comment]
I can't seem to find it now either, but I swore I saw something on it somewhere.

If you look at the poller cache for the item giving these timeouts, do you get a value back when you run it at the command line?

**Comment by anarkia1976** [comment]
Can you try to switch to spine poller?
these are my settings:
![image](https://user-images.githubusercontent.com/1744639/38015372-d90ae7e8-326b-11e8-976c-31c7302223a2.png)

8 cpus /  8G ram

This parameter is very important:
The Maximum SNMP OIDs Per SNMP Get Request

tk

**Comment by seppeel** [comment]
i switched to spine now. i will reports if that helped. maybe this could be next week.

**Comment by seppeel** [comment]
OK a short update before i have to go :-)
until now i didn't have any errors again. Assuming the spine poller won't run into the same problems as cmd.php i think there should be something to do:

* Option 1: The cmd.php poller is unstable for some reason and i will do my very best supporting you troubleshooting this problem

* Option 2: https://cacti.net/spine_info.php <- this is missleading and should be corrected.
 -> When cmd.php does not run into any errors the log looks like this:
`28.03.2018 14:25:14 - SYSTEM STATS: Time:11.4837 Method:cmd.php Processes:1 Threads:N/A Hosts:63 HostsPerProcess:63 DataSources:1575 RRDsProcessed:840 `
This is far away from 300 seconds as decribed on the info page.  The info page should describe something like "Use spine if you have more than (lets say) 1000 Data Sources" (i have 1575)

What do you think?

greetings
seppel

**Comment by netniV** [comment]
That really all comes down to many factors such as the speed of the server's cpu, disk, network, the infrastructure between your server and the hosts, and the hosts themselves.  Each of these things can cause a slowness which creeps gradually upwards.

Now, those recommendations may be wrong as they could be based on information from 10 years ago.  With the vast amount of improvements that have been occurring over that period, it may need to be updated.  However, that would need to be a gathering exercise.  In some regards, I would like to see cacti's system allowing a "Phone Home" part with the stats, whether it was cmd.php or spine and the thread/process configuration so it could be tracked.

In the short term though, we are reliant on people answering a thread on the forum to give us an idea of their usage.

**Comment by cigamit** [comment]
The real issue with cmd.php is when you have a rack of equipment that looses power, or becomes unavailable.  In those cases, the only way to go is spine.  cmd.php will just byork.  I could make it threaded and I could make it more like spine.  However, the threading is not on Windows and it is not in the core of PHP in Linux.  So, since spine is already there.  Just use spine.

**Comment by cigamit** [comment]
@seppeel, feel free to do a pull request on the Doco.  We need the help.  I'll close this for now.  Oh, and I've made some changes to the snmp library recently that did not go well.  Corrected in 1.1.38.

**Comment by seppeel** [comment]
Thats OK for me, for me its solved since i use SPINE. Unfortunately im afraid i won't find the time to edit the documentation. thanks for your snmp lib changes in 1.1.38.
